packages/smithy-http/src/smithy_http/deserializers.py

```diff
@@ -257,3 +263,72 @@ def _consume_payload(self) -> bytes:
             "Unable to read async stream. This stream must be buffered prior "
             "to creating the deserializer."
         )
+
+
+class HTTPErrorDeserializer:
+    """Binds an error response to a modelled or unknown exception."""
+
+    def __init__(
+        self,
+        payload_codec: Codec,
+        extractor: ErrorExtractor,
+        response: HTTPResponse,
+        body: "SyncStreamingBlob",
+    ) -> None:
+        """Initialize an HTTPErrorDeserializer.
+
+        :param payload_codec: The Codec to use to deserialize the payload, if present.
+        :param extractor: The error extractor to get error shape id from the response.
+        :param response: The HTTP response to read from.
+        :param body: The HTTP response body in a synchronously readable form. This is
+            necessary for async response bodies when there is no streaming member.
+        """
+        self._payload_codec = payload_codec
+        self._response = response
+        self._body = body
+        self._extractor = extractor
+        self._codec = payload_codec
+
+    def read_error(
+        self,
+        operation: APIOperation[Any, Any],
+        error_registry: TypeRegistry,
+        context: TypedProperties,
+    ) -> CallException:
+        body = self._body
+        if isinstance(body, bytearray):
+            body = bytes(body)
+        deserializer = self._payload_codec.create_deserializer(body)
+        document = deserializer.read_document(DOCUMENT)
+
+        # try to get the error shape-id from the extractor
+        error_id = self._extractor.get_error(self._response)
+
+        # if none, get it from the parsed document (e.g. '__type')
+        if error_id is None:
+            error_id = document.discriminator
```

Comment:
You're getting this document from the payload codec, but that's not necessarily gonna give you what you want. Different JSON protocols could embed this differently.